[25127] Don't reload list if query was initialized before

oliverguenther opened the pull request:
When moving from show to list, the list controller is reinitialized
and the list is reloaded through `loadQuery()`.

We can bypass this by simply checking whether the query has ever been
checked by the checksum service. Since the query observers are still
registered, changes to the query are still caught.

https://community.openproject.com/work_packages/25127/activity

oliverguenther committed:
[25127] Don't reload list if query was initialized before

When moving from show to list, the list controller is reinitialized
and the list is reloaded through `loadQuery()`.

We can bypass this by simply checking whether the query has ever been
checked by the checksum service. Since the query observers are still
registered, changes to the query are still caught.

ulferts commented:
This strangely only works after the second visit to a full screen (or maybe some other actions). But if I:
* Load a query
* click on the id link
* click on the back button

The list is reloaded when it would have to be.

oliverguenther commented:
Strange, this doesn't appear to happen locally:

![anim3](https://cloud.githubusercontent.com/assets/459462/25370206/626bdaac-2989-11e7-841e-630920dd2409.gif)

ulferts commented:
On my machine, it only happens on the first loading of the full screen:


![apr-25-2017 09-48-47](https://cloud.githubusercontent.com/assets/617519/25374115/897f9ada-299c-11e7-889f-f1825e34e297.gif)

oliverguenther commented:
This appears to be caused by the HAL resource for `/api/v3/queries/sort_bys/parent-desc` not being equal to the one stored in query. This causes the first state object in the `wpTableSortBy` to be assumed different from the query and thus, causing a list reload.

Do you have a clue why they would be different?

oliverguenther committed:
Use distinctUntilChanged to check for changes in the query states

oliverguenther committed:
Fix filters not being properly written to their $source

oliverguenther commented on frontend/app/components/api/api-v3/hal-resources/hal-resource.service.ts:
```diff
@@ -340,8 +340,9 @@ function initializeResource(halResource:HalResource) {
       halResource.$source._links[linkName] = {href: null};
-    }
-    else if (val.$link) {
+    } else if (_.isArray(val)) {
```
@ulferts This was causing `$source` not being properly updated when writing an array link. Its dubious why this never failed anywhere else.

oliverguenther commented:
PR  should be complete + your issue resolved using the `distinctUntilChanged` flow, the remaining issue was that the hal resources in the filter were both updated (and only copied in query), thus it took me a while to complete.

The broken cucumber is caused from dev which I wrote about in slack.

ulferts commented:
Code is looking good and I no longer see a reload when switching between full screen and table.

There is one issue when going back in history, however as there is an unnecessary request:

![apr-26-2017 16-51-22](https://cloud.githubusercontent.com/assets/617519/25440829/e0304906-2aa0-11e7-930e-98c1769cdfc9.gif)

The request to `api/v3/work_packages?...` is superfluous. 

The same happens, when having first loaded the page and then switching the query via the query drop down.

ulferts commented:
@oliverguenther how about we extract 3bfd51e656f1b10fdacce6c23c9e2a214e1d3cf9 from this PR and see if by that we can merge the bugfix and do the refactoring later on.

oliverguenther committed:
Move comparer and extractor functions into state value classes

This clears up what the controller has to do and allows us to double
check

1. The previous value of the state has changed
2. IF (1) is true, whether the query value for that attribute is changed

We need to check the second part additionally since the states may be
updated as a result of the query being updated in
`wpListService.updateStatesFromQuery`

The 'correct' solution here would be to create two input states, and a
dependent state that executes resulting actions.

oliverguenther commented:
@ulferts 3bfd51e was faulty since filters weren't correctly updated when switching them by hand. 

The remaining issue here was that the query is updated before the change observers fire and thus you check two things:

1. The previous value of the state has changed
2. IF (1) is true, whether the query value for that attribute is changed

We need to check the second part additionally since the states may be updated as a result of the query being updated in `wpListService.updateStatesFromQuery`

The 'correct' solution here would be to create two input states, and a dependent state that executes resulting actions.

I  believe I have resolved the issues by reverting to the double check (inside the distinctUntilChanged and with the query), however re-using the comparer function with both should result in much faster comparisons so I don't fear the two comparisons there.

ulferts commented on frontend/app/components/routing/wp-list/wp-list.controller.ts:
```diff
@@ -98,55 +105,48 @@ function WorkPackagesListController($scope:any,
+        const queryValue = stateValue.comparerFunction()(query[name]);
+        if (_.isEqual(queryValue, stateValue.extractedCompareValue)) {
+          return;
+        }
```
Now that you have pushed it this far, how about we place those three conditions in `filter` (maybe each)?

oliverguenther replied:
I did have errors sometime earlier when trying to do just that (timeline state.current being null when being accessed), but that doesn't appear to be happening any more.

ulferts commented on frontend/app/components/wp-fast-table/state/wp-table-base.service.ts:
```diff
@@ -46,10 +46,12 @@ export abstract class WorkPackageTableBaseService {
   };
 
+
+
```
???

ulferts reviewed (approved):
Very impressive @oliverguenther. 

I am willing to merge without my tiny remarks being addressed. Let me know or merge for yourself.

oliverguenther committed:
Rewrite conditions in subscribe into filters
